### ISSUE TITLE
New methods MinOrDefault and MaxOrDefault

### ISSUE THREAD

**Issue by atifaziz** [morelinq/MoreLINQ#28]
When you start playing with LINQ queries over sequences of elements (e.g. getting min / max value for enumerable source) sooner or later you will come across this one -- the `InvalidOperationException` (“Sequence contains no elements”).

The problem occurs as by default queries like `IEnumerable<T>.Min(…)` and `IEnumerable<T>.Max(…)` do not play nicely if you try to execute them on an empty sequence and just throw the exception described above. Unfortunately these methods do not have a corresponding counterpart like `Single(…) /SingleOrDefault(…)` that is smart enough to query the sequence if it is not empty or alternatively use the default value without raising an exception.

Basically you got two options now:

* Either perform the check on the enumerable sequence every time you are querying it
* OR integrate the logic in an extension method.

The second approach is much preferable so let’s add the missing link.

[http://blogs.telerik.com/manoldonev/posts/08-10-17/linq_sequence_contains_no_elements_extension_methods_to_the_rescue.aspx](http://web.archive.org/web/20110416125920/http://blogs.telerik.com/ManolDonev/Posts/08-10-17/linq_sequence_contains_no_elements_extension_methods_to_the_rescue.aspx)

---
Originally reported on Google Code with ID 28

Reported by `anton.georgiev` on 2009-09-17 08:56:23

**Comment by fsateler** [comment]
Isn't the idiomatic way to just prepend a `DefaultIfEmpty(def)` before invoking `Max` or `Min` ?

I don't think new extension methods are required here.

**Comment by MkazemAkhgary** [comment]
@fsateler whole point of linq is to reduce boilerplate code. these methods are useful and I feel its missing here...

**Comment by atifaziz** [comment]
@MkazemAkhgary That boilerplate code is precisely embodied in two methods: `DefaultIfEmpty` + `Min`/`Max`. It's just not in _one method_ and that's a reasonable trade-off considering the additional code (almost per aggregate) that would have be written, tested, documented and maintained for very little gain in the end.

**Comment by MkazemAkhgary** [comment]
> at boilerplate code is precisely embodied in two methods: DefaultIfEmpty + Min/Max

Its like saying `Cast` or `OfType` is embodied in `Select` so we don't need those.

`GroupBy` has 8 overloads. 2 of them can be covered via `Select`, why its there?

And could you explain how for example `AggregateRight`  is more useful than `Min/Max`? its strange that this library has methods for edge case usages and refuses to include more useful methods.

Imagine you had sequence of objects (reference types), and you want to get maximum of some number. here is the boilerplate code.

    var max = source.DefaultIfEmpty(null).Max(obj => obj?.foo ?? 0);

vs

    var max = source.MaxOrDefault(x => x.foo);

I think it definitely worth the effort, because its just two methods and saves bunch of characters, makes it more readable. its not hard to implement and document either.

**Comment by atifaziz** [comment]
var max = source.DefaultIfEmpty(null).Max(obj => obj?.foo ?? 0);

can be re-written more simply as:

    var max = source.Select(obj => obj.foo).DefaultIfEmpty().Max();

> Its like saying `Cast` or `OfType` is embodied in `Select` so we don't need those.

`Cast` has a slightly different raison d'être than just as a mere helper. It's used in the query syntax pattern when the source is a non-generic collection, as in `from int x in xs`. This translates to `xs.Cast<int>()`.

In a similar vein, `OfType` really exists for dealing with derived types in `IQueryable<>` sources and is there on `IEnumerable<>` for parity. [LINQ to SQL used this, for example, to support an inheritance/type hierarchy mapped over a single table](https://msdn.microsoft.com/en-us/library/bb425822.aspx#linqtosql_topic26).

> `GroupBy` has 8 overloads. 2 of them can be covered via `Select`, why its there?

Do you mean for the result projection?

> its strange that this library has methods for edge case usages…

What's an edge case to one could be common to another so it's subjective. In fact, it goes to show that MoreLINQ has had contributions that may seem uncommon and we debate & challenge this all the time; in fact, @leandromoh [did so as recently as just this week](https://github.com/morelinq/MoreLINQ/issues/380#issuecomment-364808218)!

>  …and refuses to include more useful methods.

No one is refusing. I hope you see that we are discussing here and I am listening and responding. The decision to include or exclude any proposal is not a simple and clear checklist. It's based on a number of factors that I admit (sigh) we should be documenting, at the very least those that can be made simple and clear.

**Comment by ph1l1pF** [comment]
I also would like to have those methods `MinOrDefault` and `MaxOrDefault`.
